satsense/features/texton.py

```diff
@@ -0,0 +1,143 @@
+            base_descriptors = np.append(base_descriptors, descriptors, axis=0)
+
+    # Sample {sample_size} descriptors from all descriptors
+    # (Takes random rows) and cluster these
+    print("Sampling from {}".format(str(base_descriptors.shape)))
+    X = base_descriptors[np.random.choice(
+        base_descriptors.shape[0], sample_size, replace=False), :]
+
+    mbkmeans = MiniBatchKMeans(n_clusters=n_clusters, random_state=42).fit(X)
+
+    return mbkmeans
+
+
+def create_kernels():
+    kernels = []
+    angles = 8
+    thetas = np.linspace(0, np.pi, angles)
+    for theta in thetas:
+        # theta = theta / 8. * np.pi
+        for sigma in (1, ):
+            for frequency in (0.05, ):
+                kernel = np.real(
+                    gabor_kernel(
+                        frequency, theta=theta, sigma_x=sigma, sigma_y=sigma))
+                kernels.append(kernel)
+
+    return kernels
+
+
+def compute_feats(image, kernels):
+    feats = np.zeros(image.shape + (len(kernels), ), dtype=np.double)
+    for k, kernel in enumerate(kernels):
+        filtered = ndimage.convolve(image, kernel, mode='wrap')
+        feats[:, :, k] = filtered
+
+    return feats
+
+
+class Texton(Feature):
+    def __init__(self,
+                 kmeans: MiniBatchKMeans,
+                 windows=((25, 25), ),
+                 normalized=True):
+        super(Texton, self)
+        self.windows = windows
+        self.kmeans = kmeans
+        self.feature_size = len(self.windows) * kmeans.n_clusters
+        self.normalized = normalized
+        self.kernels = create_kernels()
+        self.descriptors = None
+
+    def __call__(self, cell):
+        result = np.zeros(self.feature_size)
+        n_clusters = self.kmeans.n_clusters
+        for i, window in enumerate(self.windows):
+            win = cell.super_cell(window, padding=True)
+            start_index = i * n_clusters
+            end_index = (i + 1) * n_clusters
+            result[start_index:end_index] = self.texton(win, self.kmeans)
+        return result
+
+    def __str__(self):
+        normalized = "normalized" if self.normalized == True else "not-normalized"
+        return "Texton1-dog-{}-{}".format(str(self.windows), normalized)
+
+    def initialize(self, sat_image: SatelliteImage):
+        im_grayscale = sat_image.grayscale
+
+        descriptors = compute_feats(im_grayscale, self.kernels)
+
+        # Calculate Difference-of-Gaussian
+        dog = np.expand_dims(
+            gaussian(im_grayscale, sigma=1) - gaussian(im_grayscale, sigma=3),
+            axis=2)
+        self.descriptors = np.append(descriptors, dog, axis=2)
+
+    def texton(self, window: Cell, kmeans: MiniBatchKMeans):
+        """
+        Calculate the sift feature on the given window
+
+        Args:
+            window (nparray): A window of an image
+            maximum (int): The maximum value in the image
+        """
+        # descriptors = compute_feats(window_gray_ubyte, self.kernels)
```

Comment:
@CMFiltenborg Maybe this needs to be uncommented?

Reply:
Hi Bouwe, I changed around the code to make it more efficient, as you can compute the texton responses over the whole image in one pass...

If you call the `initialize` method before calling the feature it will work 

I added this block in my extract.py:
```
for name, feature in iteritems(features.items):
        if hasattr(feature, 'initialize'):
            feature.initialize(generator.image)
```